Parser/Token.cs

```diff
@@ -38,9 +39,37 @@ public enum E {
    // Utility function used to echo an error to the console
    public void PrintError () {
       if (Kind != ERROR) throw new Exception ("PrintError called on a non-error token");
-      Console.ForegroundColor = ConsoleColor.Yellow;
-      Console.WriteLine ($"At line {Line}, column {Column} of {Source.FileName}: {Text}");
-      Console.ResetColor ();
+      OutputEncoding = new UnicodeEncoding ();
+      var file = $"File: {Source.FileName}";
+      WriteLine (file);
+      WriteLine (new string ('\u2500', file.Length));
+
+      if (Line > 2) {
+         var (prevText, prevText1) = (Source.Lines[Line - 3], Source.Lines[Line - 2]);
+         WriteLine (string.Format ("{0,4}{1,0}{2,0}", Line - 2, "\u2502", prevText));
+         WriteLine (string.Format ("{0,4}{1,0}{2,0}", Line - 1, "\u2502", prevText1));
+      }
+      if (Line == 1)
+         WriteLine (string.Format ("{0,4}{1,0}{2,0}", Line - 1, "\u2502", Source.Lines[Line - 1]));
+
+      WriteLine (string.Format ("{0,4}{1,0}{2,0}", Line, "\u2502", Source.Lines[Line - 1]));
+
+      ForegroundColor = ConsoleColor.Yellow;
+      int padLeft = Column + 4;
+      CursorLeft = padLeft;
+      WriteLine ("^");
+      CursorLeft = Math.Min (Math.Max (0, padLeft - Text.Length / 2), WindowWidth - Text.Length);
+      WriteLine (Text);
+      ResetColor ();
+
+      var cnt = Source.Lines.Count ();
+      if (Line + 1 == cnt)
+         WriteLine (string.Format ("{0,4}{1,0}{2,0}", Line + 1, "\u2502", Source.Lines[Line + 1]));

```

Comment:
Don't use format specifiers like this. Instead rewrite this as an interpolated string.